render/pixman/renderer.c

```diff
@@ -0,0 +1,365 @@
+#include <wlr/types/wlr_matrix.h>
+#include <wlr/util/log.h>
+
+#include "render/pixman.h"
+#include "types/wlr_buffer.h"
+
+static const struct wlr_renderer_impl renderer_impl;
+
+static struct wlr_pixman_renderer *get_renderer(
+		struct wlr_renderer *wlr_renderer) {
+	assert(wlr_renderer->impl == &renderer_impl);
+	return (struct wlr_pixman_renderer *)wlr_renderer;
+}
+
+static struct wlr_pixman_buffer *get_buffer(
+		struct wlr_pixman_renderer *renderer, struct wlr_buffer *wlr_buffer) {
+	struct wlr_pixman_buffer *buffer;
+	wl_list_for_each(buffer, &renderer->buffers, link) {
+		if (buffer->buffer == wlr_buffer) {
+			return buffer;
+		}
+	}
+	return NULL;
+}
+
+static const struct wlr_texture_impl texture_impl;
+
+static struct wlr_pixman_texture *get_texture(
+		struct wlr_texture *wlr_texture) {
+	assert(wlr_texture->impl == &texture_impl);
+	return (struct wlr_pixman_texture *)wlr_texture;
+}
+
+static bool texture_is_opaque(struct wlr_texture *wlr_texture) {
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+	return texture->format->has_alpha;
+}
+
+static void texture_destroy(struct wlr_texture *wlr_texture) {
+	if (wlr_texture == NULL) {
+		return;
+	}
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+
+	pixman_image_unref(texture->image);
+	free(texture->data);
+	free(texture);
+}
+
+static const struct wlr_texture_impl texture_impl = {
+	.is_opaque = texture_is_opaque,
+	.destroy = texture_destroy,
+};
+
+struct wlr_pixman_texture *pixman_create_texture(
+		struct wlr_texture *wlr_texture, struct wlr_pixman_renderer *renderer);
+
+static void handle_destroy_buffer(struct wl_listener *listener, void *data) {
+	struct wlr_pixman_buffer *buffer =
+		wl_container_of(listener, buffer, buffer_destroy);
+	wl_list_remove(&buffer->link);
+	wl_list_remove(&buffer->buffer_destroy.link);
+
+	pixman_image_unref(buffer->image);
+
+	free(buffer);
+}
+
+static struct wlr_pixman_buffer *create_buffer(
+		struct wlr_pixman_renderer *renderer, struct wlr_buffer *wlr_buffer) {
+	struct wlr_pixman_buffer *buffer = calloc(1, sizeof(*buffer));
+	if (buffer == NULL) {
+		wlr_log_errno(WLR_ERROR, "Allocation failed");
+		return NULL;
+	}
+	buffer->buffer = wlr_buffer;
+	buffer->renderer = renderer;
+
+	uint32_t drm_format;
+	struct wlr_dmabuf_attributes dmabuf = {0};
+	struct wlr_shm_attributes shm = {0};
+	if (wlr_buffer_get_dmabuf(wlr_buffer, &dmabuf)) {
+		drm_format = dmabuf.format;
+	} else if (wlr_buffer_get_shm(wlr_buffer, &shm)) {
+		drm_format = shm.format;
+	} else {
+		goto error_buffer;
+	}
+
+	uint32_t format = get_pixman_format_from_drm(drm_format);
+
+	void *data = NULL;
+	size_t stride;
+	if (!buffer_get_data_ptr(wlr_buffer, &data, &stride)) {
+		wlr_log(WLR_ERROR, "Failed to get buffer data");
+		goto error_buffer;
+	}
+
+	wlr_log(WLR_INFO, "width %zu, height %zu", (size_t)wlr_buffer->width,
+			(size_t)wlr_buffer->height);
+	wlr_log(WLR_INFO, "data %p, stride %zu", data, stride);
```

Comment:
Can we clarify these messages a little?

- Maybe downgrade to DEBUG, since it's not really important for end-users.
- Maybe explain what's happening. For instance the GLES2 renderer logs something like "Imported 42x42 buffer as FBO"
- The value of the data pointer doesn't seem very useful.

Reply:
Ah, there's already a debug message below that reads "Created pixman buffer %dx%d". Maybe it's fine to just remove these two INFOs then?

Reply:
Oh woops that's dev output, I think we can remove those.